### PR TITLE
add notebook_dir as trait to be deprecated

### PR DESCRIPTION
**Pull request by Zsailer**
This fixes issue #35.

### CODE REVIEW

File: jupyter_server/serverapp.py [hunk 1]
```diff
@@ -1050,6 +1050,20 @@ def _update_pylab(self, change):
...
+
+    @default('notebook_dir')
+    def _default_notebook_dir(self):
+        self.log.warning(_("\n    notebook_dir is deprecated, use root_dir.\n"))
+        return self.root_dir
+
+    @validate('notebook_dir')
+    def _update_notebook_dir(self, change):
+        self.log.warning(_("\n    notebook_dir is deprecated, use root_dir.\n"))
+        self.root_dir = change['value']
```
Comment (SylvainCorlay) [inline jupyter_server/serverapp.py]: - With this behavior, there may be an issue when setting `root_dir`, and reading notebook_dir.

- Maybe an observer on `root_dir` setting the `notebook_dir` trait through set_trait (bypassing validation) would be better.

- However, we should be careful about incompatible setting of one vs the other?

Comment (Zsailer) [reply jupyter_server/serverapp.py]: Oops, that should be an `observe` decorator.

Comment (Zsailer) [reply jupyter_server/serverapp.py]: Ah, you're right. If `root_dir` changes, `notebook_dir` doesn't change.

I think setting `notebook_dir` in the `root_dir` observer should fix that. Why bypass validation? They should end up having the same validation logic right?

Comment (Zsailer) [reply jupyter_server/serverapp.py]: How about adding a 
```python
if self.notebook_dir != change['new']:
    self.notebook_dir = change['new']
```
to the `root_dir` observer, and 
```python
if self.root_dir != change['new']:
    self.root_dir = change['new']
```
to the `notebook_dir` observer?

Comment (Zsailer) [reply jupyter_server/serverapp.py]: This ensures the traits are equal.

Comment (SylvainCorlay) [reply jupyter_server/serverapp.py]: I need to do some experiments. But it is not trivial.

Comment (Zsailer) [reply jupyter_server/serverapp.py]: totally agree it's not trivial.

Comment (Zsailer) [reply jupyter_server/serverapp.py]: I added the check above if you want to test it. Just checks `notebook_dir` and `root_dir` are equal. If not, update the other trait.

**Comment by Zsailer** [comment]
@SylvainCorlay 

The pattern used in the classic notebook is to observe the deprecated trait for changes; when it is changed 1) warn the user and 2) allow it to change the new trait. Changing the new trait **does not** update the deprecated trait. 

Two examples in the classic notebook are [`base_project_url`](https://github.com/jupyter/notebook/blob/82782a370e11cddf9bc92ef87ee2244b98c24ec7/notebook/notebookapp.py#L1016) and [`server_extensions`](https://github.com/jupyter/notebook/blob/82782a370e11cddf9bc92ef87ee2244b98c24ec7/notebook/notebookapp.py#L1016).

Should I simply follow that pattern here?

```python
@observe('notebook_dir')
def _update_notebook_dir(self, change):
    self.log.warning(_('deprecation_warning'))
    self.root_dir = change['new']
```

**Comment by SylvainCorlay** [comment]
> Should I simply follow that pattern here?

I think we can do that.

**Comment by krinsman** [comment]
Will deprecating and then eventually removing `notebook_dir` make this incompatible with JupyterHub, and thus prevent `jupyter_server` from being used as a backend for `jupyterhub-singleuser`/`SingleUserNotebookApp`?

JupyterHub uses it in a pretty essential way, since that's what `c.Spawner.notebook_dir` hooks into:
https://github.com/jupyterhub/jupyterhub/blob/master/jupyterhub/spawner.py#L467
https://github.com/jupyterhub/jupyterhub/blob/master/jupyterhub/spawner.py#L930

**Comment by Zsailer** [comment]
> Will deprecating and then eventually removing notebook_dir make this incompatible with JupyterHub
@krinsman. We'll have to change this trait in JupyterHub, too.

**Comment by Zsailer** [comment]
> I think we can do that.

:+1: Merging.

**Pull request merged**